src/Functions/FunctionFormatDecimal.h

```diff
@@ -0,0 +1,306 @@
+                            "Illegal second argument for formatDecimal function: got {}, expected UInt8",
+                            arguments[1]->getName());
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t /*input_rows_count*/) const override
+    {
+        WhichDataType which_from(arguments[0].type.get());
+
+        if (which_from.isUInt8())
+            return executeType<UInt8>(arguments);
+        else if (which_from.isUInt16())
+            return executeType<UInt16>(arguments);
+        else if (which_from.isUInt32())
+            return executeType<UInt32>(arguments);
+        else if (which_from.isUInt64())
+            return executeType<UInt64>(arguments);
+        else if (which_from.isUInt128())
+            return executeType<UInt128>(arguments);
+        else if (which_from.isUInt256())
+            return executeType<UInt256>(arguments);
+        else if (which_from.isInt8())
+            return executeType<Int8>(arguments);
+        else if (which_from.isInt16())
+            return executeType<Int16>(arguments);
+        else if (which_from.isInt32())
+            return executeType<Int32>(arguments);
+        else if (which_from.isInt64())
+            return executeType<Int64>(arguments);
+        else if (which_from.isInt128())
+            return executeType<Int128>(arguments);
+        else if (which_from.isInt256())
+            return executeType<Int256>(arguments);
+        else if (which_from.isFloat32())
+            return executeType<Float32>(arguments);
+        else if (which_from.isFloat64())
+            return executeType<Float64>(arguments);
+        else if (which_from.isDecimal32())
+            return executeType<Decimal32>(arguments);
+        else if (which_from.isDecimal64())
+            return executeType<Decimal64>(arguments);
+        else if (which_from.isDecimal128())
+            return executeType<Decimal128>(arguments);
+        else if (which_from.isDecimal256())
+            return executeType<Decimal256>(arguments);
+
+        throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Illegal column {} of argument of function {}",
+                        arguments[0].column->getName(), getName());
+    }
+
+private:
+    template <typename T, std::enable_if_t<is_integer<T> || is_floating_point<T>, bool> = true>
+    ColumnPtr executeType(const ColumnsWithTypeAndName & arguments) const
+    {
+        auto result_column_string = ColumnString::create();
+        auto col_to = assert_cast<ColumnString *>(result_column_string.get());
+        ColumnString::Chars & data_to = col_to->getChars();
+        ColumnString::Offsets & offsets_to = col_to->getOffsets();
+
+        const auto * from_col = checkAndGetColumn<ColumnVector<T>>(arguments[0].column.get());
```

Comment:
`ColumnVector<T>` isn't exactly wrong. More generic would be `DataTypeNumber::ColumnType<T>`. And in l. 277: `DataTypeDecimal::ColumnType<T>`.